src/main/java/org/tron/core/services/WitnessProductBlockService.java

```diff
@@ -0,0 +1,108 @@
+package org.tron.core.services;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import org.tron.common.utils.ByteArray;
+import org.tron.core.capsule.BlockCapsule;
+
+@Slf4j
+@Service
+public class WitnessProductBlockService {
+
+  private Cache<Long, BlockCapsule> historyBlockCapsuleCache = CacheBuilder.newBuilder()
+      .initialCapacity(200).maximumSize(200).build();
+
+  private Map<String, CheatWitnessInfo> cheatWitnessInfoMap = new HashMap<>();
+
+  public static class CheatWitnessInfo {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Reduce this class from 46 to the maximum allowed 25 or externalize it in a public class. ([squid:S2972](https://sonarcloud.io/organizations/tron-zhaohong/rules#rule_key=squid%3AS2972))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=java-tron&pullRequest=1639&issues=AWalU-xJlHK86Hco_1ki&open=AWalU-xJlHK86Hco_1ki)

[ISSUE_KEY]: AWalU-xJlHK86Hco_1ki